File: app/internal/server/routes/session_history.go
package routes

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"html"
	"net/http"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/FloSch62/muxus/app/internal/api"
	"github.com/FloSch62/muxus/app/internal/history"
	"github.com/FloSch62/muxus/app/internal/server"
)

const transcriptPreviewEvents = 5_000

func configuredHistoryLocation(ctx *server.Context, settings api.SessionHistorySettings) string {
	if ctx.Config.HistoryPath != "" {
		return ctx.Config.HistoryPath
	}
	if settings.StorageLocation != "" {
		return settings.StorageLocation
	}
	if ctx.Config.DatabasePath == ":memory:" {
		return ""
	}
	return filepath.Join(filepath.Dir(ctx.Config.DatabasePath), "history")
}

func parseHistoryQuery(req *http.Request) (history.Query, error) {
	values := req.URL.Query()
	query := history.Query{
		Query:      strings.TrimSpace(values.Get("query")),
		ProfileKey: strings.TrimSpace(values.Get("profileKey")),
		Host:       strings.TrimSpace(values.Get("host")),
		Kind:       values.Get("kind"), StartedAfter: values.Get("startedAfter"),
		StartedBefore: values.Get("startedBefore"), Cursor: values.Get("cursor"),
		Limit: 50,
	}
	if len(query.Query) > 500 || len(query.ProfileKey) > 500 ||
		len(query.Host) > 500 || len(query.Cursor) > 1000 {
		return query, fmt.Errorf("session-history query is too long")
	}
	if query.ProfileKey == "" && values.Has("profileKey") {
		return query, fmt.Errorf("profileKey is required")
	}
	if query.Kind != "" && query.Kind != "ssh" && query.Kind != "local" &&
		query.Kind != "serial" && query.Kind != "telnet" {
		return query, fmt.Errorf("invalid session kind")
	}
	for _, value := range []string{query.StartedAfter, query.StartedBefore} {
		if value != "" {
			if _, err := time.Parse(time.RFC3339, value); err != nil {
				return query, fmt.Errorf("invalid ISO datetime")
			}
		}
	}
	if raw := values.Get("limit"); raw != "" {
		limit, err := strconv.Atoi(raw)
		if err != nil || limit < 1 || limit > 100 {
			return query, fmt.Errorf("limit must be between 1 and 100")
		}
		query.Limit = limit
	}
	return query, nil
}

func parsePolicyKey(req *http.Request) (string, error) {
	key := strings.TrimSpace(req.URL.Query().Get("profileKey"))
	if key == "" || len(key) > 500 {
		return "", fmt.Errorf("profileKey is required")
	}
	return key, nil
}

func RegisterSessionHistoryRoutes(r chi.Router, ctx *server.Context) {
	r.Get("/api/session-history", func(w http.ResponseWriter, req *http.Request) {
		query, err := parseHistoryQuery(req)
		if err != nil {
			writeBadRequest(w, err.Error())
			return
		}
		response, err := ctx.History.SessionHistory(query)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, response)
	})

	r.Get("/api/session-history/storage", func(w http.ResponseWriter, _ *http.Request) {
		settings, err := ctx.Database.SessionHistorySettings()
		if err != nil {
			writeInternalError(w, err)
			return
		}
		status, err := ctx.History.StorageStatus(configuredHistoryLocation(ctx, settings))
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, status)
	})
	r.Put("/api/session-history/storage", func(w http.ResponseWriter, req *http.Request) {
		body, ok := readJSONBody(w, req, defaultBodyLimit)
		if !ok {
			return
		}
		var input api.SessionHistorySettingsInput
		if err := json.Unmarshal(body, &input); err != nil {
			writeBadRequest(w, "invalid session-history settings")
			return
		}
		settings, err := ctx.Database.SaveSessionHistorySettings(input)
		if err != nil {
			writeBadRequest(w, err.Error())
			return
		}
		if err := ctx.History.UpdateSettings(settings); err != nil {
			writeInternalError(w, err)
			return
		}
		status, err := ctx.History.StorageStatus(configuredHistoryLocation(ctx, settings))
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, status)
	})

	r.Get("/api/session-history/policy", func(w http.ResponseWriter, req *http.Request) {
		key, err := parsePolicyKey(req)
		if err != nil {
			writeBadRequest(w, err.Error())
			return
		}
		policy, err := ctx.Database.SessionLoggingPolicy(key)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, policy)
	})
	r.Put("/api/session-history/policy", func(w http.ResponseWriter, req *http.Request) {
		key, err := parsePolicyKey(req)
		if err != nil {
			writeBadRequest(w, err.Error())
			return
		}
		body, ok := readJSONBody(w, req, defaultBodyLimit)
		if !ok {
			return
		}
		var input api.SessionLoggingPolicyInput
		if err := json.Unmarshal(body, &input); err != nil {
			writeBadRequest(w, "invalid session logging policy")
			return
		}
		policy, err := ctx.Database.SaveSessionLoggingPolicy(key, input)
		if err != nil {
			writeBadRequest(w, err.Error())
			return
		}
		server.WriteJSON(w, http.StatusOK, policy)
	})
	r.Delete("/api/session-history/policy", func(w http.ResponseWriter, req *http.Request) {
		key, err := parsePolicyKey(req)
		if err != nil {
			writeBadRequest(w, err.Error())
			return
		}
		deleted, err := ctx.Database.DeleteSessionLoggingPolicy(key)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, deletedResponse{Deleted: deleted})
	})

	r.Get("/api/session-history/{id}", func(w http.ResponseWriter, req *http.Request) {
		limit := transcriptPreviewEvents
		session, err := ctx.History.SessionLog(chi.URLParam(req, "id"), &limit)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		if session == nil {
			writeNotFound(w, "session log not found")
			return
		}
		server.WriteJSON(w, http.StatusOK, session)
	})
	r.Get("/api/session-history/{id}/raw", func(w http.ResponseWriter, req *http.Request) {
		zero := 0
		session, err := ctx.History.SessionLog(chi.URLParam(req, "id"), &zero)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		events, err := ctx.History.RawSessionLogEvents(chi.URLParam(req, "id"))
		if err != nil {
			writeInternalError(w, err)
			return
		}
		if session == nil || events == nil {
			writeNotFound(w, "session log not found")
			return
		}
		w.Header().Set("Content-Type", "application/x-ndjson")
		w.Header().Set(
			"Content-Disposition",
			`attachment; filename="`+exportSlug(session.Title)+`.muxlog"`,
		)
		encoder := json.NewEncoder(w)
		for _, event := range events {
			_ = encoder.Encode(map[string]any{
				"sequence": event.Sequence, "recordedAt": event.RecordedAt,
				"elapsedMs": event.ElapsedMs, "direction": event.Direction,
				"encoding": "base64", "data": base64.StdEncoding.EncodeToString(event.Raw),
			})
		}
	})
	r.Get("/api/session-history/{id}/clean", func(w http.ResponseWriter, req *http.Request) {
		session, err := ctx.History.SessionLog(chi.URLParam(req, "id"), nil)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		if session == nil {
			writeNotFound(w, "session log not found")
			return
		}
		var transcript strings.Builder
		for _, event := range session.Events {
			transcript.WriteString(event.Text)
		}
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		w.Header().Set(
			"Content-Disposition",
			`attachment; filename="`+exportSlug(session.Title)+`-clean.txt"`,
		)
		_, _ = w.Write([]byte(transcript.String()))
	})
	r.Get("/api/session-history/{id}/replay.html", func(w http.ResponseWriter, req *http.Request) {
		session, err := ctx.History.SessionLog(chi.URLParam(req, "id"), nil)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		if session == nil {
			writeNotFound(w, "session log not found")
			return
		}
		sendReplay(w, session)
	})
	r.Put("/api/session-history/{id}/pin", func(w http.ResponseWriter, req *http.Request) {
		body, ok := readJSONBody(w, req, defaultBodyLimit)
		if !ok {
			return
		}
		var input struct {
			Pinned *bool `json:"pinned"`
		}
		if json.Unmarshal(body, &input) != nil || input.Pinned == nil {
			writeBadRequest(w, "pinned is required")
			return
		}
		updated, err := ctx.History.SetPinned(chi.URLParam(req, "id"), *input.Pinned)
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, map[string]bool{"updated": updated})
	})
	r.Delete("/api/session-history/{id}", func(w http.ResponseWriter, req *http.Request) {
		deleted, err := ctx.History.DeleteSession(chi.URLParam(req, "id"))
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, deletedResponse{Deleted: deleted})
	})
}

var slugNonAlpha = regexp.MustCompile(`[^a-z0-9]+`)

func exportSlug(title string) string {
	value := strings.Trim(slugNonAlpha.ReplaceAllString(strings.ToLower(title), "-"), "-")
	if len(value) > 48 {
		value = value[:48]
	}
	if value == "" {
		return "session"
	}
	return value
}

func sendReplay(w http.ResponseWriter, session *api.SessionLogDetail) {
	type replayEvent struct {
		Time      int64  `json:"t"`
		Direction string `json:"d"`
		Text      string `json:"x"`
	}
	events := make([]replayEvent, len(session.Events))
	for index, event := range session.Events {
		events[index] = replayEvent{
			Time: event.ElapsedMs, Direction: event.Direction, Text: event.Text,
		}
	}
	raw, _ := json.Marshal(events)
	eventsJSON := strings.NewReplacer(
		"<", `\u003c`, "\u2028", `\u2028`, "\u2029", `\u2029`,
	).Replace(string(raw))
	title := html.EscapeString(session.Title)
	host := html.EscapeString(session.Host)
	started := html.EscapeString(session.StartedAt)
	page := `<!doctype html><html lang="en"><head><meta charset="utf-8">` +
		`<meta name="viewport" content="width=device-width,initial-scale=1">` +
		`<title>` + title + ` — Muxus replay</title><style>` +
		`:root{color-scheme:dark;background:#0c1117;color:#d7dee7;font:14px system-ui,sans-serif}` +
		`*{box-sizing:border-box}body{margin:0;height:100vh;display:flex;flex-direction:column}` +
		`header{padding:14px 18px;background:#121a23;border-bottom:1px solid #293442}` +
		`h1{font-size:16px;margin:0 0 4px}.meta{color:#8fa0b3;font-size:12px}` +
		`.controls{display:flex;align-items:center;gap:10px;padding:10px 18px;background:#0f171f}` +
		`button,select{color:inherit;background:#1b2733;border:1px solid #3a4959;border-radius:5px;padding:6px 10px}` +
		`input{flex:1}.time{min-width:110px;text-align:right}pre{flex:1;margin:0;padding:18px;overflow:auto;white-space:pre-wrap;font:13px/1.5 monospace}` +
		`.input{color:#8ec7ff}.system{color:#8391a2;font-style:italic}</style></head><body>` +
		`<header><h1>` + title + `</h1><div class="meta">` + host + ` · started ` + started +
		` · ` + strconv.FormatInt(session.EventCount, 10) + ` retained events</div></header>` +
		`<div class="controls"><button id="play">Play</button><button id="restart">Restart</button>` +
		`<label>Speed <select id="speed"><option value=".5">0.5×</option><option value="1" selected>1×</option><option value="2">2×</option><option value="5">5×</option><option value="20">20×</option></select></label>` +
		`<input id="seek" type="range" min="0" max="0" value="0"><span class="time" id="time">00:00 / 00:00</span></div><pre id="screen"></pre>` +
		`<script>const events=` + eventsJSON + `,duration=events.length?events.at(-1).t:0;` +
		`const screen=document.querySelector('#screen'),play=document.querySelector('#play'),seek=document.querySelector('#seek'),clock=document.querySelector('#time'),speed=document.querySelector('#speed');seek.max=duration;let position=0,index=0,running=false,last=0;const fmt=ms=>{const s=Math.floor(ms/1000),m=Math.floor(s/60);return String(m).padStart(2,'0')+':'+String(s%60).padStart(2,'0')};function append(e){const x=document.createElement('span');x.className=e.d;x.textContent=e.x;screen.append(x)}function render(to){screen.textContent='';index=0;while(index<events.length&&events[index].t<=to)append(events[index++]);position=to;seek.value=to;clock.textContent=fmt(to)+' / '+fmt(duration)}function frame(now){if(!running)return;position=Math.min(duration,position+(now-last)*Number(speed.value));last=now;while(index<events.length&&events[index].t<=position)append(events[index++]);seek.value=position;clock.textContent=fmt(position)+' / '+fmt(duration);if(position>=duration){running=false;play.textContent='Play';return}requestAnimationFrame(frame)}play.onclick=()=>{running=!running;play.textContent=running?'Pause':'Play';last=performance.now();if(running)requestAnimationFrame(frame)};restart.onclick=()=>{running=false;play.textContent='Play';render(0)};seek.oninput=()=>render(Number(seek.value));render(0)</script></body></html>`
	w.Header().Set("Content-Type", "text/html")
	w.Header().Set(
		"Content-Disposition",
		`attachment; filename="`+exportSlug(session.Title)+`-replay.html"`,
	)
	_, _ = w.Write([]byte(page))
}

File: server/src/routes/session-history.ts
import type { FastifyInstance, FastifyReply } from 'fastify';
import path from 'node:path';
import { z } from 'zod';
import type {
  SessionHistoryResponse,
  SessionHistoryStorageStatus,
  SessionLogDetail,
  SessionLoggingPolicy,
} from '@muxus/shared';
import type { AppContext } from '../app.js';
import { defaultHistoryRoot } from '../session-logging/history-store.js';
import { HttpProblem, sendError } from '../util/errors.js';

const TRANSCRIPT_PREVIEW_EVENTS = 5_000;

const historyQuerySchema = z.object({
  query: z.string().trim().max(500).optional(),
  profileKey: z.string().trim().min(1).max(500).optional(),
  host: z.string().trim().max(500).optional(),
  kind: z.enum(['ssh', 'local', 'serial', 'telnet']).optional(),
  startedAfter: z.iso.datetime().optional(),
  startedBefore: z.iso.datetime().optional(),
  limit: z.coerce.number().int().min(1).max(100).default(50),
  cursor: z.string().max(1000).optional(),
});

const policyKeySchema = z.object({
  profileKey: z.string().trim().min(1).max(500),
});

const policySchema = z.object({
  enabled: z.boolean(),
  captureInput: z.boolean(),
  maxPartBytes: z.number().int().min(64 * 1024).max(1024 * 1024 * 1024),
  maxParts: z.number().int().min(1).max(1000),
});

const historySettingsSchema = z.object({
  storageLocation: z.string().trim().max(4096).optional(),
  maxTotalBytes: z.number().int().min(64 * 1024 * 1024),
  minFreeBytes: z.number().int().min(0),
  minFreePercent: z.number().min(0).max(100),
  maxAgeDays: z.number().int().min(1).optional(),
}).superRefine((value, ctx) => {
  if (value.storageLocation && !path.isAbsolute(value.storageLocation)) {
    ctx.addIssue({
      code: 'custom',
      path: ['storageLocation'],
      message: 'storageLocation must be an absolute path',
    });
  } else if (
    value.storageLocation &&
    path.parse(path.resolve(value.storageLocation)).root ===
      path.resolve(value.storageLocation)
  ) {
    ctx.addIssue({
      code: 'custom',
      path: ['storageLocation'],
      message: 'storageLocation cannot be the filesystem root',
    });
  }
});

const pinSchema = z.object({ pinned: z.boolean() });

const detailQuerySchema = z.object({
  query: z.string().trim().max(500).optional(),
});

export function registerSessionHistoryRoutes(
  app: FastifyInstance,
  ctx: AppContext,
): void {
  app.get('/api/session-history', async (req): Promise<SessionHistoryResponse> => {
    const parsed = historyQuerySchema.parse(req.query);
    return ctx.history.sessionHistory(parsed);
  });

  app.get(
    '/api/session-history/storage',
    async (): Promise<SessionHistoryStorageStatus> => {
      const settings = ctx.database.sessionHistorySettings();
      return ctx.history.storageStatus(configuredStorageLocation(ctx, settings));
    },
  );

  app.put(
    '/api/session-history/storage',
    async (req, reply): Promise<SessionHistoryStorageStatus | void> => {
      try {
        const settings = ctx.database.saveSessionHistorySettings(
          historySettingsSchema.parse(req.body),
        );
        await ctx.history.updateSettings(settings);
        return ctx.history.storageStatus(configuredStorageLocation(ctx, settings));
      } catch (err) {
        return sendError(reply, asRouteError(err));
      }
    },
  );

  app.get('/api/session-history/policy', (req): SessionLoggingPolicy => {
    const { profileKey } = policyKeySchema.parse(req.query);
    return ctx.database.sessionLoggingPolicy(profileKey);
  });

  app.put(
    '/api/session-history/policy',
    async (req, reply): Promise<SessionLoggingPolicy | void> => {
      try {
        const { profileKey } = policyKeySchema.parse(req.query);
        const input = policySchema.parse(req.body);
        return ctx.database.saveSessionLoggingPolicy(profileKey, input);
      } catch (err) {
        return sendError(reply, asRouteError(err));
      }
    },
  );

  app.delete('/api/session-history/policy', (req): { deleted: boolean } => {
    const { profileKey } = policyKeySchema.parse(req.query);
    return { deleted: ctx.database.deleteSessionLoggingPolicy(profileKey) };
  });

  app.get(
    '/api/session-history/:id',
    async (req, reply): Promise<SessionLogDetail | void> => {
      const { id } = req.params as { id: string };
      const { query } = detailQuerySchema.parse(req.query);
      const session = await ctx.history.sessionLog(
        id,
        TRANSCRIPT_PREVIEW_EVENTS,
        query || undefined,
      );
      if (!session) {
        await reply.code(404).send({ message: 'session log not found' });
        return;
      }
      return session;
    },
  );

  app.get('/api/session-history/:id/raw', async (req, reply) => {
    const { id } = req.params as { id: string };
    const [session, events] = await Promise.all([
      ctx.history.sessionLog(id, 0),
      ctx.history.rawSessionLogEvents(id),
    ]);
    if (!session || !events) {
      await reply.code(404).send({ message: 'session log not found' });
      return;
    }
    const body = events
      .map((event) =>
        JSON.stringify({
          sequence: event.sequence,
          recordedAt: event.recordedAt,
          elapsedMs: event.elapsedMs,
          direction: event.direction,
          encoding: 'base64',
          data: event.raw.toString('base64'),
        }),
      )
      .join('\n');
    return reply
      .type('application/x-ndjson')
      .header(
        'content-disposition',
        `attachment; filename="${exportSlug(session.title)}.muxlog"`,
      )
      .send(body ? `${body}\n` : '');
  });

  app.get('/api/session-history/:id/clean', async (req, reply) => {
    const { id } = req.params as { id: string };
    const session = await ctx.history.sessionLog(id);
    if (!session) {
      await reply.code(404).send({ message: 'session log not found' });
      return;
    }
    const transcript = session.events.map((event) => event.text).join('');
    return reply
      .type('text/plain; charset=utf-8')
      .header(
        'content-disposition',
        `attachment; filename="${exportSlug(session.title)}-clean.txt"`,
      )
      .send(transcript);
  });

  app.get('/api/session-history/:id/replay.html', async (req, reply) => {
    const { id } = req.params as { id: string };
    const session = await ctx.history.sessionLog(id);
    if (!session) {
      await reply.code(404).send({ message: 'session log not found' });
      return;
    }
    return sendReplay(reply, session);
  });

  app.put('/api/session-history/:id/pin', async (req): Promise<{ updated: boolean }> => {
    const { id } = req.params as { id: string };
    const { pinned } = pinSchema.parse(req.body);
    return { updated: await ctx.history.setPinned(id, pinned) };
  });

  app.delete('/api/session-history/:id', async (req): Promise<{ deleted: boolean }> => {
    const { id } = req.params as { id: string };
    return { deleted: await ctx.history.deleteSession(id) };
  });

  app.setErrorHandler((err, _req, reply) => {
    if (err instanceof z.ZodError) {
      return sendError(
        reply,
        new HttpProblem(400, err.issues[0]?.message ?? 'invalid session-history request'),
      );
    }
    return sendError(reply, err);
  });
}

function sendReplay(reply: FastifyReply, session: SessionLogDetail) {
  const eventsJson = JSON.stringify(
    session.events.map((event) => ({
      t: event.elapsedMs,
      d: event.direction,
      x: event.text,
    })),
  )
    .replaceAll('<', '\\u003c')
    .replaceAll('\u2028', '\\u2028')
    .replaceAll('\u2029', '\\u2029');
  const title = escapeHtml(session.title);
  const host = escapeHtml(session.host);
  const started = escapeHtml(session.startedAt);
  const html = `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>${title} — Muxus replay</title>
<style>
:root{color-scheme:dark;background:#0c1117;color:#d7dee7;font:14px Inter,system-ui,sans-serif}
*{box-sizing:border-box}body{margin:0;height:100vh;display:flex;flex-direction:column}
header{padding:14px 18px;background:#121a23;border-bottom:1px solid #293442}
h1{font-size:16px;margin:0 0 4px}.meta{color:#8fa0b3;font-size:12px}
.controls{display:flex;align-items:center;gap:10px;padding:10px 18px;background:#0f171f;border-bottom:1px solid #293442}
button,select{color:inherit;background:#1b2733;border:1px solid #3a4959;border-radius:5px;padding:6px 10px}
input[type=range]{flex:1}.time{font-variant-numeric:tabular-nums;min-width:110px;text-align:right;color:#aab8c6}
pre{flex:1;margin:0;padding:18px;overflow:auto;white-space:pre-wrap;word-break:break-word;font:13px/1.5 ui-monospace,SFMono-Regular,Consolas,monospace}
.input{color:#8ec7ff}.system{color:#8391a2;font-style:italic}
</style>
</head>
<body>
<header><h1>${title}</h1><div class="meta">${host} · started ${started} · ${session.eventCount} retained events</div></header>
<div class="controls">
  <button id="play" type="button">Play</button>
  <button id="restart" type="button">Restart</button>
  <label>Speed <select id="speed"><option value="0.5">0.5×</option><option value="1" selected>1×</option><option value="2">2×</option><option value="5">5×</option><option value="20">20×</option></select></label>
  <input id="seek" type="range" min="0" max="0" value="0" aria-label="Replay position">
  <span class="time" id="time">00:00 / 00:00</span>
</div>
<pre id="screen" aria-live="off"></pre>
<script>
const events=${eventsJson};
const duration=events.length?events[events.length-1].t:0;
const screen=document.querySelector('#screen'),play=document.querySelector('#play'),seek=document.querySelector('#seek'),time=document.querySelector('#time'),speed=document.querySelector('#speed');
seek.max=String(duration);let position=0,index=0,running=false,last=0;
const fmt=ms=>{const s=Math.floor(ms/1000),m=Math.floor(s/60);return String(m).padStart(2,'0')+':'+String(s%60).padStart(2,'0')};
function append(e){const span=document.createElement('span');span.className=e.d;span.textContent=e.x;screen.append(span)}
function render(to){screen.textContent='';index=0;while(index<events.length&&events[index].t<=to)append(events[index++]);screen.scrollTop=screen.scrollHeight;position=to;seek.value=String(to);time.textContent=fmt(to)+' / '+fmt(duration)}
function frame(now){if(!running)return;const delta=(now-last)*Number(speed.value);last=now;position=Math.min(duration,position+delta);while(index<events.length&&events[index].t<=position)append(events[index++]);screen.scrollTop=screen.scrollHeight;seek.value=String(position);time.textContent=fmt(position)+' / '+fmt(duration);if(position>=duration){running=false;play.textContent='Play';return}requestAnimationFrame(frame)}
play.onclick=()=>{running=!running;play.textContent=running?'Pause':'Play';if(running){last=performance.now();requestAnimationFrame(frame)}};
document.querySelector('#restart').onclick=()=>{running=false;play.textContent='Play';render(0)};
seek.oninput=()=>render(Number(seek.value));render(0);
</script>
</body>
</html>`;
  return reply
    .type('text/html')
    .header(
      'content-disposition',
      `attachment; filename="${exportSlug(session.title)}-replay.html"`,
    )
    .send(html);
}

function escapeHtml(value: string): string {
  return value
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&#39;');
}

function exportSlug(title: string): string {
  return (
    title
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/^-+|-+$/g, '')
      .slice(0, 48) || 'session'
  );
}

function asRouteError(err: unknown): unknown {
  if (!(err instanceof z.ZodError)) return err;
  return new HttpProblem(
    400,
    err.issues[0]?.message ?? 'invalid session-history request',
  );
}

function configuredStorageLocation(
  ctx: AppContext,
  settings: { storageLocation?: string },
): string | undefined {
  return (
    ctx.config.historyPath ??
    settings.storageLocation ??
    defaultHistoryRoot(ctx.config.databasePath)
  );
}
